File: 77-combinations/combinations.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        output = [] 

        def findCombinations(i, curr):
            if i > n + 1:
                return

            if len(curr) == k:
                output.append(curr.copy())
                return

            curr.append(i)

            findCombinations(i + 1, curr)

            curr.pop()

            findCombinations(i + 1, curr)
        
        findCombinations(1, [])
        return output

        

File: 77-combinations/combinations.cpp
class Solution {
    vector<vector<int>> output;
    vector<int> curr;

    void findCombinations(int start, int n, int k) {
        if (k == 0) {
            output.push_back(curr);
            return;
        }

        if (start > n) {
            return;
        }

        for (int i = start; i <= n; i++) {
            curr.push_back(i);
            findCombinations(i+1, n, k-1);
            curr.pop_back();
        }
    }

public:
    vector<vector<int>> combine(int n, int k) {
        findCombinations(1, n, k);
        return output;
    }
};